Cloud diagnostics for Lauer et al. (2022)

axel-lauer opened the pull request:
This PR implements the new and updated cloud diagnostics (and documentation) used in Lauer et al. (2022), J. Climate.

New recipes: recipes/clouds/recipe_lauer22jclim_fig*.yml

These recipes need the following new/updated datasets:

* CERES-EBAF Ed4.1: #2752
* CALIPSO-ICECLOUD: #2753
* CLOUDSAT-L2: #2754

In addition, this extension of the ERA5 CMORizer is required:

* ERA5 extension: https://github.com/ESMValGroup/ESMValCore/pull/1850

Documentation: https://esmvaltool--2750.org.readthedocs.build/en/2750/recipes/recipe_clouds.html

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [x] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [x] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [x] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [x] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [x] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [x] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

remi-kazeroni reviewed (changes requested):
Many thanks for all your work and contribution @axel-lauer! There's quite some science done with these recipes! This PR looks in very good shape to me. Most recipes run fine and quickly. A few recipes need more compute resources or a quick fix w.r.t the OBS data (see below). Apart from a few minor typos, my main questions/comments would be:

- Documentation: in some cases, I had trouble figuring out which diagnostics of the recipe produce the plots showcased in the documentation. It seemed to me that some generated plots look different from the documentation (more/less models or features missing/added) but it could just be that I'm not used enough to these plots. Would it be possible to take a look at the docs and if needed update the plots and/or the captions?

- Naming of the plots: Some recipes generate a large number of plots per model. I guess this allows users to perform detailed evaluation of cloud properties for many models at a time. But since the png files do not contain the name of the model used, I think it a bit difficult to browse the plots. Would it be possible to add the name of the model to the filename? Or maybe at least in cases where a multi-model mean is shown?

- OBS in recipes: for the `recipe_lauer22jclim_fig2_taylor_amip.yml`, one needs to move the `CLOUDSAT-L2` data from Tier2 to Tier3. There is also an issue with the ERA5 data used in `recipe_lauer22jclim_fig3-4_zonal.yml` which is not using the on-the-fly cmorization.

- Already public recipes: In this PR, 3 recipes were moved to a `clouds` directory but not changed. I used the comparison tool between this branch and the main one for these 3 recipes. This reported 0 change in all 3 cases.

- Issues with the masking: I believe the plots won't be affected by that once the new Core is released.

- Computational resources (just for the records): I think 2 recipes can't be run unless using a compute node with 512GB of memory on Levante. That is `recipe_lauer22jclim_fig3-4_zonal.yml` and `recipe_lauer22jclim_fig5_lifrac.yml`

- Diagnostic bug: I think there is an issue with the `clouds_lifrac_scatter_postproc.ncl` diagnostic script which can't be run unless the path to the root directory does not contain any capital letter. A log file for this can be seen in `/work/bd0854/b309192/runs/tests/clouds_updates_PR/esmvaltool_output/recipe_lauer22jclim_fig5_lifrac_20221123_163640/run/postproc/cmip5cmip6/log.txt`.

It would be great if the scientific review could be done soon so that the plots could be checked by a more expert person than me. I'd be happy to generate a website that shows the output for all the recipes of this PR if needed.

remi-kazeroni commented:
Here is a website to visualize the output from the set of recipes of this PR: [overview page](https://esmvaltool.dkrz.de/shared/esmvaltool/cloud_updates_PR2750/) or [debug page](https://esmvaltool.dkrz.de/shared/esmvaltool/cloud_updates_PR2750/debug.html). Note that the preprocessed files are not available on the website, there is not enough disk space available for that. The actual runs can be found at `/work/bd0854/b309192/runs/tests/clouds_updates_PR/esmvaltool_output` (only take a look at the successful ones with a `index.html` file, the others failed due to insufficient resources). Hope that helps the scientific review 👍

axel-lauer commented on doc/sphinx/source/recipes/recipe_clouds.rst:
```diff
@@ -279,15 +527,17 @@ Example plots
    Taylor diagram showing the 20-yr annual average performance of CMIP5 models
-   for total cloud fraction as compared to MODIS satellite observations.
+   for total cloud fraction as compared to MODIS satellite observations. Produced
+   with recipe_lauer13jclim.yml.
```
Good point. I updated all example figures in the documentation with up-to-date versions.

axel-lauer commented on doc/sphinx/source/recipes/recipe_clouds.rst:
```diff
@@ -303,12 +553,13 @@ Example plots
    individual CMIP5 models (thin gray lines) and the multi-model mean (thick
-   red line). Similar to Figure 9.5 of Flato et al. (2013).
+   red line). Similar to Figure 9.5 of `Flato et al., 2013`_. Produced
+   with recipe_clouds_ipcc.yml.
```
Thanks for spotting this. The masking issue indeed affected this diagnostic. The problem with mask_fillvalues is fixed now (https://github.com/ESMValGroup/ESMValCore/pull/1823) and the new example figure looks fine.

axel-lauer commented on doc/sphinx/source/recipes/recipe_clouds.rst:
```diff
@@ -317,5 +568,49 @@ Example plots
+   2-dimensional distribution of average total cloud cover (clt) binned by sea
+   surface temperature (SST, x-axis) and vertical velocity at 500 hPa (ω\ :sub:`500`, y-axis)
+   averaged over 20 years and all grid cells over the ocean. Produced with
+   recipe_lauer22jclim_fig8_dyn.yml.
```
The panels are optional and turned on/off with the recipe settings (parameter "sidepanels"). I updated the example figure now not showing any side panels to avoid confusion.

axel-lauer commented on esmvaltool/references/lauer22jclim.bibtex:
```diff
@@ -0,0 +1,13 @@
+@article{lauer22jclim,
```
The paper is available as "early online release". I can add the missing fields once the paper has been published in one of the journal's issues. I hope that's fine...

axel-lauer commented:
Thanks @remi-kazeroni for doing the technical review!!! I think I addressed all your points (see below) with the exeption of these two points:

> * Naming of the plots: Some recipes generate a large number of plots per model. I guess this allows users to perform detailed evaluation of cloud properties for many models at a time. But since the png files do not contain the name of the model used, I think it a bit difficult to browse the plots. Would it be possible to add the name of the model to the filename? Or maybe at least in cases where a multi-model mean is shown?
> * Computational resources (just for the records): I think 2 recipes can't be run unless using a compute node with 512GB of memory on Levante. That is `recipe_lauer22jclim_fig3-4_zonal.yml` and `recipe_lauer22jclim_fig5_lifrac.yml`

Renaming the output plots would be quite a bit of work. The automatic numbering is done by NCL for file formats (such as .png) that do not support multi-page documents (such as .ps or .pdf). I started using multi-page plots to avoid "cluttering" and easier handling, so from my point of view doing all this work for improved support of single-page file formats is probably not really worth it and I would propose to leave it as is for now...

You are right about the high memory consumption. recipe_lauer22jclim_fig3-4_zonal.yml uses up to 160 GB (1 task), recipe_lauer22jclim_fig5_lifrac.yml up to 171 GB (1 task). I think that's the many 3-d variables being preprocessed. Looks like another incentive to make more preprocessor functions lazy...

axel-lauer commented on doc/sphinx/source/recipes/recipe_clouds.rst:
```diff
@@ -6,45 +6,85 @@ Clouds
+   (cl) and water vapor path (prw) from CMIP6 models in comparison to CMIP5 results
+   and satellite observations. Whereever possible, the diagnostics use
+   multi-observational products as reference datatsets. The recipe family
+   reproduces all figures from `Lauer et al. (2022)`_: maps of the geographical
```
I updated the link (and rererence). Thanks for spotting this!

remi-kazeroni reviewed (approved):
Thanks a lot for addressing my comments and your answers @axel-lauer! This all looks good to me from a technical point of view. I reran the recipes affected by the latest changes and all worked as expected. I could also more easily find in the run directories the plots showcased in the documentation, thanks for updating that. I added the label "requires new ESMValCore release" but strictly speaking that is only true for one of the recipes of this PR (and also for those affected by the wrong behaviour of the masking of 0s). Once this has been scientifically reviewed and approved, it can be merged 👍

LisaBock commented on esmvaltool/diag_scripts/clouds/clouds_taylor_show.ncl:
```diff
@@ -0,0 +1,38 @@
+; CLOUDS_LIFRAC_SCATTER_POSTPROC
+; ############################################################################
+; Author: Axel Lauer (DLR, Germany)
+; ############################################################################
+; Description
+;   Read ice fraction from clouds_lifrac_scatter.ncl and plot results for
+;   cmip5/cmip6 model pairs defined below.
+;
+; Modification history
+;   20211113-lauer_axel: written.
+;
+; ############################################################################
```
Is this file still needed?
The inpath is here hardcoded...

axel-lauer replied:
Ups... this file should already have been deleted a long time ago. No idea why this slipped through. I removed the file now.

LisaBock reviewed (changes requested):
@axel-lauer Thanks a lot for your great scientific work with the ESMValTool!
The figures are looking all fine when comparing them to the published once. So, in general, the PR is in a very could shape.

Two points:
- I agree with Rémi that it is difficult to find specific plots in the figure files which do not contain the model or obs name in the filename. But I understand that it is maybe not worth to put the effort in to change the filenames...
- Figure 9-11c: I could find the barplot style of the figure but not the pdf line style as it is in the publication. Is there a setting missing in the recipe?

And thanks @remi-kazeroni for providing all the results and plots to me!

axel-lauer commented on esmvaltool/diag_scripts/clouds/clouds.ncl:
```diff
@@ -197,21 +248,111 @@ begin
+      delete(ymean)
+    end do
+    delete(sigma2)
+    ; note: we are using dim_avg_n_Warp so missing values are ignored
+    ;       when averaging
+    ref_std = sqrt(dim_avg_n(sig_tmp, 0))
```
The functions dim_avg_n and dim_avg_n_Wrap both ignore missing values, the only difference is whether the metadata is preserved (dim_avg_n_Wrap) or not (dim_avg_n). I updated the comments in clouds.ncl to avoid confusion.

axel-lauer commented:
Thanks @LisaBock for taking a look!

> Two points:
> 
>     * I agree with Rémi that it is difficult to find specific plots in the figure files which do not contain the model or obs name in the filename. But I understand that it is maybe not worth to put the effort in to change the filenames...

I guess the easiest option (for now) is to use a file format supporting multi-page documents (.ps, .pdf). I never understood why the default setting in config-user.yml is .png...

>     * Figure 9-11c: I could find the barplot style of the figure but not the pdf line style as it is in the publication. Is there a setting missing in the recipe?

That was a problem with .png output and has been fixed with this commit https://github.com/ESMValGroup/ESMValTool/pull/2750/commits/0eea904c7414ba95ee2d15ac2e423e34194bd2b3 I would guess you were still looking at output created before I introduced this fix...

LisaBock reviewed (approved):
Thanks @axel-lauer ! It looks fine for me now.
